src/wallet/coinselection.cpp

```diff
@@ -341,3 +341,30 @@ CAmount OutputGroup::GetSelectionAmount() const
 {
     return m_subtract_fee_outputs ? m_value : effective_value;
 }
+
+CAmount GetSelectionWaste(const std::set<CInputCoin>& inputs, const CAmount change_cost, const CAmount target)
+{
+    // If the result is empty, then selection failed. Don't use this one, so set waste to very high
+    if (inputs.empty()) {
+        return MAX_MONEY;
+    }
+
+    // Always consider the cost of spending an input now vs in the future.
+    CAmount waste = 0;
+    CAmount selected_value = 0;
+    for (const CInputCoin& coin : inputs) {
+        waste += coin.m_fee - coin.m_long_term_fee;
+        selected_value += coin.txout.nValue;
+    }
+
+    // Consider the cost of making change and spending it in the future
+    // If we aren't making change, the caller should've set change_cost to 0
+    waste += change_cost;
+
+    if (change_cost == 0) {
+        // When we are not making change (change_cost == 0), consider the excess we are throwing away to fees
+        waste += selected_value - target;
+    }
```

Comment:
This is not just the excess, but `selected_value - target` will also include all input fees paid for the input set. However, we counted the input fees already above in the comparison with the `long_term_fee`.

I think this should be `selected_value - txweight*feerate - sum(recipient outputs)`, or alternatively additionally deduct `sum(coin.m_fee)` for each input for the code as currently written.

A few good tests for the waste metric would be:
- Transaction with change has excess of 0
- Transaction with feerate of 10 sat/vB has waste that is exactly `cost_of_change`+excess (one of which would be zero in the case of having change and the other in the case of not having change)
- Transaction with same count of inputs has higher waster at higher feerate than at lower feerate.

Reply:
Changed to calculated `selected_value` with `coin.effective_value`. This will ensure the fee is subtracted. Also renamed it to `selected_effective_value`.

Reply:
Ah, very elegant!